stdlib/src/utils/span.mojo

```diff
@@ -0,0 +1,222 @@
+# ===----------------------------------------------------------------------=== #
+# Copyright (c) 2024, Modular Inc. All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+
+"""Implements the Span type.
+
+You can import these APIs from the `utils.span` module. For example:
+
+```mojo
+from utils import Span
+```
+"""
+
+from . import InlineArray
+
+
+@value
+struct _SpanIter[
+    T: CollectionElement,
+    is_mutable: __mlir_type.`i1`,
+    lifetime: AnyLifetime[is_mutable].type,
+    forward: Bool = True,
+]:
+    """Iterator for Span.
+
+    Parameters:
+        T: The type of the elements in the span.
+        is_mutable: Whether the reference to the span is mutable.
+        lifetime: The lifetime of the Span.
+        forward: The iteration direction. `False` is backwards.
+    """
+
+    var index: Int
+    var src: Span[T, is_mutable, lifetime]
+
+    @always_inline
+    fn __iter__(self) -> Self:
+        return self
+
+    @always_inline
+    fn __next__(
+        inout self,
+    ) -> Reference[T, is_mutable, lifetime]:
+        @parameter
+        if forward:
+            self.index += 1
+            return self.src._refitem__(self.index - 1)
+        else:
+            self.index -= 1
+            return self.src._refitem__(self.index)
+
+    @always_inline
+    fn __len__(self) -> Int:
+        @parameter
+        if forward:
+            return len(self.src) - self.index
+        else:
+            return self.index
+
+
+@value
+struct Span[
+    T: CollectionElement,
+    is_mutable: __mlir_type.i1,
+    lifetime: AnyLifetime[is_mutable].type,
+]:
+    """A non owning view of contiguous data.
+
+    Parameters:
+        T: The type of the elements in the span.
+        is_mutable: Whether the span is mutable.
+        lifetime: The lifetime of the Span.
+    """
+
+    var data: UnsafePointer[T]
+    var len: Int
+
+    @always_inline
+    fn __init__(inout self, *, data: UnsafePointer[T], len: Int):
+        """Unsafe construction from a pointer and length.
+
+        Args:
+            data: The underlying pointer of the span.
+            len: The length of the view.
+        """
+        self.data = data
+        self.len = len
+
+    @always_inline
```

Comment:
Note that we probably want to get rid of this constructor in favor of returning it from `__getitem__(self: Reference[Self, _, _], slice: Slice) -> Span[T, _ , _]`

Reply:
That's totally fine to do in a follow-up though IMO.  Let's introduce the new type and then apply its use throughout the library.